### ISSUE TITLE
how to understand the MPC in samtools stats output

### ISSUE THREAD

**Issue by litun-fkby** [samtools/samtools#1954]
Hi，
I want to know the columns mean about the MPC in the samtools stats output, it shows:
![image](https://github.com/samtools/samtools/assets/82378273/1f6da5bb-74ac-4f80-9f67-1e33f90aaf96)

The first columns is the cycle number ,second is the number of N's and the rest is the number of mismatches , but it also see  columns correspond to qualities , it's make me confused ,can you give me a detail for the each columns?

Thanks!

**Comment by pd3** [comment]
Not sure how to explain it better, will this make it more clear:

We count the number of mismatches in aligned reads at each position, across all reads. Rows correspond to read cycles (i.e. position within the read) and columns to qualities, the values are the number of bases for that read cycle and quality. However, there are two extra columns. The first is the read cycle number, the second is the number of N's (for those we do not collect base qualities).

If still unclear, check also the code
https://github.com/samtools/samtools/blob/a4c73e4d41670957be6c68ff964a0310f58be9ab/stats.c#L1617-L1628

**Comment by litun-fkby** [comment]
you mean the value in 1st row and 3rd column mean's the mismatch number of 1st read cycle and the base quality 0 , as follows:
![image](https://github.com/samtools/samtools/assets/82378273/1d6e9884-d01b-441a-9629-d04693679ca7)

can samtools stats output the total base in each cycle , and samtools output the cycle value consider  the read1 , read2 , forward alignment and reverse alignment ?

**Comment by pd3** [comment]
> you mean the value in 1st row and 3rd column mean's the mismatch number of 1st read cycle and the base quality 0

Yes, there are no mismatching bases with BQ=0, at any cycle; these would be turned into N's by the base caller I am assuming.


> can samtools stats output the total base in each cycle , and samtools output the cycle value consider the read1 , read2 , forward alignment and reverse alignment ?

Not in one go. But you can run the `stats` command twice, each time pre-filtering the reads you want.

**Comment by litun-fkby** [comment]
_Not in one go. But you can run the stats command twice, each time pre-filtering the reads you want._
I can pre-filtering the read1 and read2 to get the MPC result，but how to get the total base per cycle if i want to get the mismatch rate per cycle fastly.

**Comment by pd3** [comment]
Well, you sum up numbers in the columns.

**Comment by jkbonfield** [comment]
> > you mean the value in 1st row and 3rd column mean's the mismatch number of 1st read cycle and the base quality 0
> 
> Yes, there are no mismatching bases with BQ=0, at any cycle; these would be turned into N's by the base caller I am assuming.

You'd like to think so, but my experience of writing FASTQ compressors says otherwise.  There are both non-N bases with quality 0 and also N bases with quality non-zero.  It's silly and breaks the meaning of the Phred scale, but sadly there are many buggy base callers too.

It's not something that stats is concerned with I assume though as it's just presenting the raw stats obviously.

**Comment by pd3** [comment]
That statement was related to the emphasized column in the attached screenshot, there all BQ=0 have zero counts. I am sure not all bams are like that.

**Comment by litun-fkby** [comment]
_Well, you sum up numbers in the columns._
Total base mean the all the base in per read cycle,which include the no mismatch base and the mismatch base , not the sum of the mismatch base .

**Comment by pd3** [comment]
You can get that information from the other stats. For example, there is the ACGT content by cycle, or qualities per cycle.

**Comment by jkbonfield** [comment]
Testing this I don't understand what a mismatch is either.  I'm assuming "N" isn't a mismatch?  Feels a bit odd, but I guess it's not an explicit call that differs to the reference, as it's a non-call.

Really this all needs documenting.  MPC doesn't exist at all in the man page and didn't exist in any stats file I produced until I hunted through the code to figure out that it only appears when specifying a reference file.

**Comment by litun-fkby** [comment]
_You can get that information from the other stats. For example, there is the ACGT content by cycle, or qualities per cycle._
thank you!
@jkbonfield the MPC documenting not exist , maybe the samtools stats documenting is not timely update.
